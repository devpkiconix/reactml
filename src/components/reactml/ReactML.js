import React from 'react';
import {
    __, defaultTo, path as pathGet,
    curry,
} from 'ramda';
import mapValues from 'lodash/mapValues';
import dependencies from './dependencies';
import renderLib from './render';
import defaultTagFactory from './materialUiTagFactory';

const renderer = curry(renderLib.render)(dependencies);

const { connect, withStyles } = dependencies;

const defaultToEmpty = defaultTo({})

const state2PropsMaker = (compName, props) => {
    const state2propsPath = ['spec', 'components', compName, 'state-to-props'];
    const stateNodeName = props.spec.state.stateNodeName;

    const oState2pathSpec = defaultToEmpty(
        pathGet(state2propsPath, props)
    );
    return (state) => {
        const imPathGet =
            (dotted) => state.reactml.getIn([stateNodeName].concat(dotted.split('.')));
        const mapped = mapValues(oState2pathSpec, imPathGet);
        // console.log(state.reactml.toJS(), "mapped", mapped);
        return mapped;
    };
};

export const ReactML = (props) => {
    const stateNodeName = props.stateNodeName;
    const compName = props.component;
    const stylesPath = ['spec', 'components', compName, 'styles'];
    const viewNodePath = ['spec', 'components', compName, 'view']
    const mapStateToProps = state2PropsMaker(compName, props);
    const actionExtractor = () => props.actionLib;
    const compProps = {
        tagFactory: props.tagFactory || defaultTagFactory,
        root: pathGet(viewNodePath, props),
        // ...mapStateToProps,
        stateNodeName,
    };
    return React.createElement(
        connect(mapStateToProps, actionExtractor())(
            withStyles(
                defaultToEmpty(
                    pathGet(stylesPath, props)
                ))(
                    renderer
                )
        ),
        compProps);
};

