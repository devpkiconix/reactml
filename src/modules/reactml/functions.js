import React from 'react';
import {
    __, curry, path as pathGet, map as RMap, omit, keys
} from 'ramda';
import { isString, isObject } from 'ramda-adjunct';

import { normalizeNode } from './normalize';

import { fromYaml } from './util';

const isLeaf = (node) => {
    // const res = isString(node) || !node.children ||
    //     (node.children && node.children.length == 0);
    const res = !isObject(node);
    // console.log("isLeaf? ", node, res);
    return res;
}
const mapPropName2Value = curry((tagGetter, rootProps, dottedName) => {
    // console.log(`interpreting prop name: [${dottedName}]`);
    let value = dottedName; // default
    if (isString(dottedName) && dottedName.startsWith('...')) {
        // triple tag is a react component
        value = tagGetter({ tag: dottedName.substr(3) });
    } else if (isString(dottedName) && dottedName.startsWith('..')) {
        // Double dot is a function
        value = rootProps[dottedName.substr(2)];
    } else if (isString(dottedName) && dottedName[0] == '.') {
        // single dot is a prop
        value = pathGet(dottedName.substr(1).split('.'), rootProps);
    }
    return value;
});

// Given a reactml node, determine the react
// class/function (i.e. JSX tag)
const mapNode2Tag = curry((tagFactory, node) => {
    // console.log("node", node)
    if (isString(node)) {
        return node;
    }
    if (!node.tag) {
        throw new Error("Invalid configuration. Specify a tag name")
    }
    return tagFactory[node.tag] || node.tag;
});

const maybeParse = (data) => (isString(data)) ? fromYaml(data) : data;

const _mapPropsTree = (propGetter, tagGetter, node) => {
    const mappedProps = RMap(propGetter, node.props || {})
    return {
        ...node,
        props: mappedProps,
        content: node.content ? propGetter(node.content) : null,
        children: node.children.map(child => _mapPropsTree(propGetter, tagGetter, child))
    };
};

const sansProps = omit(['props', 'tag', 'content']);
const node2children = (node) => {
    let children = node.content ? [node.content] : node.children;
    // console.log("Children for", node, 'are ', children);
    return children;
}

const traverse = (basicRender, tagGetter, propGetter) => {
    const nodeProcessor = (node, key = 0) => {
        // console.log('nodeProcessor:', node, key)
        let
            children = null,
            mappedChildren = null,
            mappedProps = null;
        if (isLeaf(node)) {
            children = [];
            mappedProps = null;
        } else {
            let children = node2children(node);
            mappedChildren = children ? children.map(nodeProcessor) : null;
            mappedProps = RMap(propGetter, node.props || {});
            mappedProps.key = key;
        }
        let rendered = basicRender(
            tagGetter, propGetter,
            mappedProps, mappedChildren, node);
        return rendered;
    };
    return nodeProcessor;
}


export const renderTree = (propGetter, tagFactory) => {
    const tagGetter = mapNode2Tag(tagFactory);
    const treeWalker = traverse(basicRenderReact, tagGetter, propGetter);
    return (tree) => treeWalker(tree, 0);
}

const codegenTree = (propGetter, _) => (tree) => {
    const tagGetter = (node) => node.tag;

    return traverse(basicRenderCodegen, tagGetter, propGetter)(tree);
}


const basicRenderReact = (tagGetter, propGetter, mappedProps, mappedChildren, node) => {
    if (isString(node)) {
        return propGetter(node);
    }
    return React.createElement(
        tagGetter(node), mappedProps, mappedChildren
    );
};

const basicRenderCodegen = (tagGetter, propGetter, mappedProps, mappedChildren, node) => {
    let tag = tagGetter(node);
    if (isString(node)) {
        return propGetter(node);
    }
    return `
<${tag} ${propsStr}>
${childrenStr}
</${tag}>
    `;
}

// Exported functions for reuse and/or testing
export default {
    isLeaf,
    maybeParse, mapNode2Tag, mapPropName2Value, traverse,
    renderTree, codegenTree, node2children, normalizeNode,
};

