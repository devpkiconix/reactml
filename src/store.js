import { createStore, applyMiddleware, compose } from 'redux'
import { connectRouter, routerMiddleware } from 'connected-react-router'
import thunk from 'redux-thunk'
import logger from 'redux-logger'
import createHistory from 'history/createBrowserHistory'
import rootReducer from './modules'


export const history = createHistory()

const initialState = {}
const enhancers = []
const middleware = [thunk, routerMiddleware(history)]

if (process.env.NODE_ENV === 'development') {
  const devToolsExtension = window.__REDUX_DEVTOOLS_EXTENSION__

  middleware.push(logger)
  if (typeof devToolsExtension === 'function') {
    enhancers.push(devToolsExtension())
  }
}

const composedEnhancers = compose(
  applyMiddleware(...middleware),
  ...enhancers
)


function makeStore() {
  const store = createStore(
    connectRouter(history)(rootReducer),
    initialState,
    composedEnhancers
  );
  return store;
}

export default makeStore;