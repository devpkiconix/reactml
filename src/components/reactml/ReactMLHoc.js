import React from 'react';
import { connect } from 'react-redux';
import { ReactML } from '../../components/reactml/ReactML';

const ReactmlHoc = (spec, actionLib, tagFactory, stateNodeName, view) => {
    class View extends React.Component {
        componentDidMount() {
            // If initial state has been provided by YAML
            // and not already been initialized, then
            // generate a redux action to init state
            if (spec.state.initial &&
                this.props.stateInitialized === false) {

                this.props.dispatch({
                    type: 'REACTML_INIT', initial: {
                        ...spec.state.initial,
                        spec, actionLib, tagFactory,
                        stateNodeName,
                    },
                    stateNodeName,
                });
            }
        }

        render() {
            return (<ReactML
                tagFactory={tagFactory}
                stateNodeName={stateNodeName}
                spec={spec} component={view}
                actionLib={actionLib}
            />);
        }
    }

    const mapStateToProps = (state) => {
        const reactml = state.reactml;
        let stateInitialized = false;
        // If the state has already been initialized,
        // set a flag so we don't have to init on every HOC
        if (reactml.get(stateNodeName)) {
            stateInitialized = true;
        }
        return {stateInitialized};
    };
    return connect(mapStateToProps)(View);
}


export default ReactmlHoc;
