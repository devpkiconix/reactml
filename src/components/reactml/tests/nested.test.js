import React from 'react';
import { Provider } from 'react-redux'
import { ReactML } from '../ReactML';
import ReactMLArrayMapper from '../ArrayMapper';
import ReactMLHoc from '../ReactMLHoc';
import renderer from 'react-test-renderer';
import store from './store';
import functions from '../../../modules/reactml/functions';

const { maybeParse } = functions;

const tagFactory = {};
const lib = {};

describe("Reactml rendering of nested components", () => {
	it('auto-add to tag factory', () => {
        const spec = {
            state: {
                stateNodeName: 'compstate1'
            },
            components: {
                page1: {
                    view: {
                        tag: 'div',
                        children: [
                            { tag: 'span', content: 'hello world' },
                            { tag: 'LowLevelComp' },
                            { tag: 'AnotherLowLevelComp' },
                        ]
                    }
                },
                LowLevelComp: {
                	view: { tag: 'span', content: 'low' },
                },
                AnotherLowLevelComp: {
                	view: { tag: 'span', content: 'low too' },
                },
            },
        };
        const tree = renderer
            .create(
                <Provider store={store}>
                    <ReactML
                        tagFactory={tagFactory}
                        stateNodeName="compstate1"
                        spec={spec} component="page1"
                        actionLib={lib}
                    />
                </Provider>)
            .toJSON();
        expect(tree).toMatchSnapshot();	});
});