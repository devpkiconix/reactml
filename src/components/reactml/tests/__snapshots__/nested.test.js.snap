// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reactml rendering of nested components auto-add to tag factory 1`] = `
<div>
  <span>
    hello world
  </span>
  <span>
    low
  </span>
  <span>
    low too
  </span>
</div>
`;
